Image registration

seanbudd opened the pull request:
## Description
[Tell us about your new feature, improvement, or fix! If relevant, please supplement the PR with images.]


## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]


## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

seanbudd commented:
> @seanbudd can we add the latest changes from Shachar here? I'd also like him to be committing as well.

done. Shachar is able to push to this branch as well, he just wanted me to implement the doc tests so we can work on the project simultaneously. I think (hope) from now on we will only be working on this branch - no separate file. Fun note - looks like the GIF stuff will be available soon https://github.com/sphinx-gallery/sphinx-gallery/issues/150#issuecomment-422153749

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+        image = pyramid_reduce(image, sigma=2/3)
+        pyramid[level] = (image)
+
+    return pyramid
```
You can use `reversed(pyramid_gaussian(image, levels=6))` instead of writing this function.

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+    return pyramid
+
+
+def _mse(img1, img2):
```
This probably belongs in `measure.metrics` as `mean_squared_error`

seanbudd replied:
working on this

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+
+def _mse(img1, img2):
+    """
+    Finds the error between the overlapping area of two images
```
Read PEP257 "docstring conventions". Here, move this line to be inline with the """ and reword to:

Find the mean squared error between the overlapping area of two images.

seanbudd replied:
working on this

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+    Parameters
+    ----------
+    img1 : (M, N) ndarray
+        Input image used for reference
```
Use whatever convention is used in `measure.metrics`, I think it's `im_reference` and `im_test`?

seanbudd replied:
working on this

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+    err: int
+        Error in the form of the squared difference between pixels
+    """
+    dim0 = min(img1.shape[0],img2.shape[0])
```
PEP8, space after comma

jni replied:
Actually, this whole function needs to be updated to support n-dimensional images:

```
limits = tuple(slice(None, min(s, t)) for s, t in zip(img1.shape, img2.shape))
return np.sum((img1[limits] - img2[limits])**2)
```

seanbudd replied:
@leGIT-bot do you want this or shall I do it

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+        Error in the form of the squared difference between pixels
+    """
+    transformation = p_to_matrix(param)
+    transformed = warp(target_image, transformation, order=3)
```
We don't want to use `warp` here because it uses "backwards" coordinates. Use `ndi.affine_transform` instead

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+    """
+    
+    r, tc, tr = param
+    return SimilarityTransform(rotation=r, translation=(tc, tr))
```
You're not returning a matrix here, you're returning a SimilarityTransform!

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+    """
+    
+    m = matrix.params
+    return (np.arccos(m[0][0])*180/np.pi, m[0][2], m[1][2])
```
a) use np.rad2deg, and
b) why are we using degs???

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+    return (np.arccos(m[0][0])*180/np.pi, m[0][2], m[1][2])
+
+
+def register(reference, target, *, cost=_cost_mse, nlevels=7, method='Powell', iter_callback=lambda img, p: None):
```
Call this register_affine.

jni commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,143 @@
+    p = np.zeros(3)
+
+    for n, (ref, tgt) in zip(levels, image_pairs):
+        p[1] *= 2
```
`p` is a vector, so you can use `p[1:3] *= 2`. 

Also, since you mentioned that you have to decrease the downscaling factor in some cases, this is a good thing to parametrise.

GenevieveBuckley commented on doc/examples/transform/plot_register_align.py:
```diff
@@ -0,0 +1,73 @@
+import numpy as np
+from skimage import measure
+
+rows=6
```
PEP8 style recommends spaces on either side of the equals assignment operator - can you change this to match? Thanks!

https://www.python.org/dev/peps/pep-0008/ 
"Always surround these binary operators with a single space on either side: assignment (=), augmented assignment (+=, -= etc.), comparisons (==, <, >, !=, <>, <=, >=, in, not in, is, is not), Booleans (and, or, not)."

GenevieveBuckley replied:
(Might be worth taking another quick check over the rest of the code with autopep8)

GenevieveBuckley commented on doc/examples/transform/plot_register_align.py:
```diff
@@ -0,0 +1,73 @@
+ax[5].set_xticks(np.arange(x/5, x, x/5), minor=True)
+ax[5].set_yticks(np.arange(y/5, y, y/5), minor=True)
+ax[5].grid(which='minor', color='w', linestyle='-', linewidth=1)
+plt.show()
```
I'm getting a little message here saying there is no newline at the end of this file - could you add one?

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,89 @@
+    Converts a transformation of the first 2 rows of a 3x3 transformation matrix
+    to the full matrix
+
+    Parameters
```
Your docstring calls your inputs/outputs "param" and "matrix" but in the function code they are called "p" and "out".
1. Please replace these with more meaningful variable names.
2. Variable names should match in the docstring and the function code.

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,89 @@
+
+def register_affine(reference, target, *, cost=compare_mse, nlevels=None,
+                    iter_callback=lambda img, matrix: None):
+
```
Super important to have a docstring here - this is the function accessible to users, so it should have great documentation (including a short example).

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,89 @@
+    return out
+
+
+def _matrix_to_p(matrix):
```
I really like the symmetry between the function names _matrix_to_p and _p_to_matrix.
Can we make it clearer what "p" is here? Ideally we could choose something more immediately obvious.

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,89 @@
+
+
+def _matrix_to_p(matrix):
+    """
```
The docstring summary should fit on a single line. 
This one (and the one before it) are slightly too long. You can shorten the summary to one line, then have a blank line, and additional detail/explanation after that (remember to put another blank line between that and your parameters, etc).
https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,89 @@
+
+
+def register_affine(reference, target, *, cost=compare_mse, nlevels=None,
+                    iter_callback=lambda img, matrix: None):
```
I'm not sure how I feel about lambda functions in a function call. I assume @jni recommended this?

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,89 @@
+            return cost(ref, transformed)
+
+        res = minimize(_cost, p, method='Powell')
+        p = res.x
```
You might consider renaming 'res', just because it's only one letter off 'ref' & could be a little confusing for someone skimming though this section of the code later on.

GenevieveBuckley commented on doc/examples/transform/plot_register_align.py:
```diff
@@ -0,0 +1,73 @@
+from skimage.data import camera
```
Can you organise and group your import statements here according to the PEP8 recommndations? Thanks!

That is:
1. Standard library imports
2. Third party library imports
3. Imports from scikit-image
(with a blank line between each group).
https://www.python.org/dev/peps/pep-0008/#imports

seanbudd commented:
Hi @GenevieveBuckley 

I've cleaned up skimage/transform/registration.py some more and I think it's ready to go. 
The example and unittests are still TODO but I might leave them until after the report/exams if that's ok.
Thanks for the suggestions on how to do the unittests, I wouldn't have thought of using allclose to check registration. 

I've renamed p/param with argmin as that appears to be a [mathematical convention](https://en.wikipedia.org/wiki/Mathematical_optimization#Optimal_input_arguments)

I'm not sure what would be better to replace with the function parameter than the lambda, as the lambda clearly shows parameters for the function. It's also used for the morphological_geodesic_active_contour in skimage/segmentation/morphsnakes.py
If there's a suggestion for a better convention I'm happy to change it - I mainly just copied the idea from that function and I love lambdas. 

Cheers,
Sean

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,129 @@
+    matrix : array
+        A transformation matrix used to obtain a new image.
+        ndi.affine_transform(target, matrix) will align your target image.
+
```
Nice docstring! Can you please add a minimal example section to it?

Maybe something like this? (Check it works first, obviously).

```
Example
--------
>>> reference_image = camera()
>>> matrix_transform = [
        [np.cos(0.12), -np.sin(0.12), 0.2],
        [np.sin(0.12),  np.cos(0.12), 0.1],
        [0,             0,            1.0]]
>>> target_image = ndi.affine_transform(reference_image, matrix_transform)
>>> output_image = register_affine(reference_image, target_image)
```

GenevieveBuckley commented on skimage/transform/registration.py:
```diff
@@ -0,0 +1,129 @@
+    """
+    Returns a matrix which registers the target image to the reference image
+
+    Uses image registration techniques, and only optimises with affine transformations.
```
Be *really* specific about what "techniques" means here, or reword the sentence if you mean to say it's image registration using affine transformations.

GenevieveBuckley commented on doc/examples/transform/plot_register_align.py:
```diff
@@ -0,0 +1,71 @@
+callback_arr = []
+callback = lambda i, m: callback_arr.append((i,m))
+r = 0.12
+out = np.empty((3, 3))
```
I'm not sure 'out' is a very helpful variable name here, particularly since it's not actually the output from your function.

It might be more helpful to have:
out -> matrix_transform
cam -> reference_image
start -> target_image
trans -> output_image / registered_image / something like that.

GenevieveBuckley commented on doc/examples/transform/plot_register_align.py:
```diff
@@ -0,0 +1,71 @@
+r = 0.12
+out = np.empty((3, 3))
+
+out[0][0] = np.cos(r)
```
Maybe write the whole array in a single line, rather than over 9 or 10 lines.

```
matrix_transform = [
    [np.cos(r), -np.sin(r), 0.2],
    [np.sin(r),  np.cos(r), 0.1],
    [0,          0,         1.0]]
```

GenevieveBuckley commented:
> @GenevieveBuckley 👋 from 🇩🇪 =)
> 
> I'm definitely going to bed now but yes, the idea for this PR was to be able to comment before it was polished enough for a direct PR to skimage. After this next round of comments I think that time will have come!

Good to know. I'm expecting we'll have a short pause while there are other more pressing deadlines for this project, and then @seanbudd  and @leGIT-bot are pretty close!
